File: java/src/com/ctre/phoenix/Drive/Mecanum.java
package com.ctre.phoenix.drive;
//package com.ctre.phoenix.Drive;
//
//import com.ctre.phoenix.Mechanical.*;
//import com.ctre.phoenix.motorcontrol.*;
//import com.ctre.phoenix.Drive.Styles;
//
//public class Mecanum implements IDrivetrain
//{
//	Gearbox _1;
//	Gearbox _2;
//	Gearbox _3;
//	Gearbox _4;
//	
//	public Mecanum(IMotorController m1, IMotorController m2, IMotorController m3, IMotorController m4)
//	{
//		GroupMotorControllers.Register(m1);
//		
//		Gearbox temp1 = new Gearbox(m1);
//		Gearbox temp2 = new Gearbox(m2);
//		Gearbox temp3 = new Gearbox(m3);
//		Gearbox temp4 = new Gearbox(m4);
//		
//		_1 = temp1;
//		_2 = temp2;
//		_3 = temp3;
//		_4 = temp4;
//	}
//	
//	public void set(Styles.Basic mode, double forward, double turn)
//	{
//		Drive(mode, forward, 0, turn);
//	}
//	
//	public void set(Styles.Basic mode, double forward, double strafe, double turn)
//	{
//		Drive(mode, forward, strafe, turn);
//	}
//	
//	public void SetVoltageRampRate(double rampRate)
//    {
//        _1.SetVoltageRampRate(rampRate);
//        _2.SetVoltageRampRate(rampRate);
//        _3.SetVoltageRampRate(rampRate);
//        _4.SetVoltageRampRate(rampRate);
//    }
//
//    public void SetVoltageCompensationRampRate(double rampRate)
//    {
//        _1.SetVoltageCompensationRampRate(rampRate);
//        _2.SetVoltageCompensationRampRate(rampRate);
//        _3.SetVoltageCompensationRampRate(rampRate);
//        _4.SetVoltageCompensationRampRate(rampRate);
//    }
//
//    public void ConfigPeakPercentOutputVoltage(double forwardVoltage, double reverseVoltage)
//    {
//        _1.ConfigPeakOutputVoltage(forwardVoltage, reverseVoltage);
//        _2.ConfigPeakOutputVoltage(forwardVoltage, reverseVoltage);
//        _3.ConfigPeakOutputVoltage(forwardVoltage, reverseVoltage);
//        _4.ConfigPeakOutputVoltage(forwardVoltage, reverseVoltage);
//    }
//
//    public void ConfigNominalPercentOutputVoltage(double forwardVoltage, double reverseVoltage)
//    {
//        _1.ConfigNominalOutputVoltage(forwardVoltage, reverseVoltage);
//        _2.ConfigNominalOutputVoltage(forwardVoltage, reverseVoltage);
//        _3.ConfigNominalOutputVoltage(forwardVoltage, reverseVoltage);
//        _4.ConfigNominalOutputVoltage(forwardVoltage, reverseVoltage);
//    } 
//    
//    private void Drive(Styles.Basic mode, double forward, double strafe, double turn)
//    {
//        double leftFrnt_throt = (forward + strafe + turn); // left front moves positive for forward, strafe-right, turn-right
//        double leftRear_throt = (forward - strafe + turn); // left rear moves positive for forward, strafe-left, turn-right
//        double rghtFrnt_throt = (forward - strafe - turn); // right front moves positive for forward, strafe-left, turn-left
//        double rghtRear_throt = (forward + strafe - turn); // right rear moves positive for forward, strafe-right, turn-left
//
//        /* Set control mode */
//        if(mode == Styles.Basic.PercentOutput)
//        {
//            _1.SetControlMode(ControlMode.BasicControlMode.kPercentVbus);
//            _2.SetControlMode(ControlMode.BasicControlMode.kPercentVbus);
//            _3.SetControlMode(ControlMode.BasicControlMode.kPercentVbus);
//            _4.SetControlMode(ControlMode.BasicControlMode.kPercentVbus);
//        }
//        else if(mode == Styles.Basic.Voltage)
//        {
//            _1.SetControlMode(ControlMode.BasicControlMode.kVoltage);
//            _2.SetControlMode(ControlMode.BasicControlMode.kVoltage);
//            _3.SetControlMode(ControlMode.BasicControlMode.kVoltage);
//            _4.SetControlMode(ControlMode.BasicControlMode.kVoltage);
//        }
//
//        /* Set motors */
//        _1.set(leftFrnt_throt);
//        _2.set(leftRear_throt);
//        _3.set(rghtFrnt_throt);
//        _4.set(rghtRear_throt);
//    }
//}

File: cpp/src/Drive/Mecanum.cpp
//#include "ctre/phoenix/MotorControl/DeviceCatalog.h"
//#include "ctre/phoenix/Drive/Mecanum.h"
//#include "HAL/DriverStation.h"
//#include <cmath>
//
//namespace ctre { namespace drive {
//
//Mecanum::Mecanum(ctre::motorcontrol::IMotorController *m1, ctre::motorcontrol::IMotorController *m2,
//		ctre::motorcontrol::IMotorController *m3, ctre::motorcontrol::IMotorController *m4)
//{
//	ctre::Mechanical::Gearbox *temp1 = new ctre::Mechanical::Gearbox(m1);
//	ctre::Mechanical::Gearbox *temp2 = new ctre::Mechanical::Gearbox(m2);
//	ctre::Mechanical::Gearbox *temp3 = new ctre::Mechanical::Gearbox(m3);
//	ctre::Mechanical::Gearbox *temp4 = new ctre::Mechanical::Gearbox(m4);
//
//	_1 = temp1;
//	_2 = temp2;
//	_3 = temp3;
//	_4 = temp4;
//}
//Mecanum::Mecanum(ctre::Mechanical::Gearbox *m1, ctre::Mechanical::Gearbox *m2,
//		ctre::Mechanical::Gearbox *m3, ctre::Mechanical::Gearbox *m4)
//{
//	_1 = m1;
//	_2 = m2;
//	_3 = m3;
//	_4 = m4;
//}
//void Mecanum::Set(Styles::Basic mode, float forward, float turn){
//	Drive(mode, forward, 0, turn);
//}
//void Mecanum::Set(Styles::Basic mode, float forward, float strafe, float turn){
//	Drive(mode, forward, strafe, turn);
//}
//void Mecanum::SetVoltageRampRate(float rampRate){
//	_1->SetVoltageRampRate(rampRate);
//	_2->SetVoltageRampRate(rampRate);
//	_3->SetVoltageRampRate(rampRate);
//	_4->SetVoltageRampRate(rampRate);
//}
//void Mecanum::SetVoltageCompensationRampRate(float rampRate){
//    _1->SetVoltageCompensationRate(rampRate);
//    _2->SetVoltageCompensationRate(rampRate);
//    _3->SetVoltageCompensationRate(rampRate);
//    _4->SetVoltageCompensationRate(rampRate);
//}
//void Mecanum::ConfigPeakPercentOutputVoltage(float forwardVoltage, float reverseVoltage){
//    _1->ConfigPeakOutputVoltage(forwardVoltage, reverseVoltage);
//    _2->ConfigPeakOutputVoltage(forwardVoltage, reverseVoltage);
//    _3->ConfigPeakOutputVoltage(forwardVoltage, reverseVoltage);
//    _4->ConfigPeakOutputVoltage(forwardVoltage, reverseVoltage);
//}
//void Mecanum::ConfigNominalPercentOutputVoltage(float forwardVoltage, float reverseVoltage){
//    _1->ConfigNominalOutputVoltage(forwardVoltage, reverseVoltage);
//    _2->ConfigNominalOutputVoltage(forwardVoltage, reverseVoltage);
//    _3->ConfigNominalOutputVoltage(forwardVoltage, reverseVoltage);
//    _4->ConfigNominalOutputVoltage(forwardVoltage, reverseVoltage);
//}
//void Mecanum::Drive(Styles::Basic mode, float forward, float strafe, float turn){float leftFrnt_throt = (forward + strafe + turn); // left front moves positive for forward, strafe-right, turn-right
//	float leftRear_throt = (forward - strafe + turn); // left rear moves positive for forward, strafe-left, turn-right
//	float rghtFrnt_throt = (forward - strafe - turn); // right front moves positive for forward, strafe-left, turn-left
//	float rghtRear_throt = (forward + strafe - turn); // right rear moves positive for forward, strafe-right, turn-left
//
//	/* Set control mode */
//	if(mode == Styles::Basic::PercentOutputBasic)
//	{
//		_1->SetControlMode(ctre::motorcontrol::ControlMode::BasicControlMode::kPercentVbusBasic);
//		_2->SetControlMode(ctre::motorcontrol::ControlMode::BasicControlMode::kPercentVbusBasic);
//		_3->SetControlMode(ctre::motorcontrol::ControlMode::BasicControlMode::kPercentVbusBasic);
//		_4->SetControlMode(ctre::motorcontrol::ControlMode::BasicControlMode::kPercentVbusBasic);
//	}
//	else if(mode == Styles::Basic::VoltageBasic)
//	{
//		_1->SetControlMode(ctre::motorcontrol::ControlMode::BasicControlMode::kVoltageBasic);
//		_2->SetControlMode(ctre::motorcontrol::ControlMode::BasicControlMode::kVoltageBasic);
//		_3->SetControlMode(ctre::motorcontrol::ControlMode::BasicControlMode::kVoltageBasic);
//		_4->SetControlMode(ctre::motorcontrol::ControlMode::BasicControlMode::kVoltageBasic);
//	}
//
//	/* Set motors */
//	_1->Set(leftFrnt_throt);
//	_2->Set(leftRear_throt);
//	_3->Set(rghtFrnt_throt);
//	_4->Set(rghtRear_throt);
//}
//
//}}
